doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,902 @@
+                ieee802154_rx_info_t *info);
+```
+
+### PHY Operations
+```c
+    /**
+     * @brief Perform Standalone Clear Channel Assessment
+     *
+     * This function performs blocking CCA to check if the channel is clear.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     *
+     * @return true if channel is clear.
+     * @return false if channel is busy.
+     */
+    bool (*cca)(ieee802154_dev_t *dev);
+
+    /**
+     * @brief Set the threshold for the Energy Detection (first mode of CCA)
+     *
+     * @pre the device is on
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] threshold the threshold in dBm.
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_cca_threshold)(ieee802154_dev_t *dev, int8_t threshold);
+
+    /**
+     * @brief Set CCA mode
+     *
+     * All radios MUST at least implement the first CCA mode (ED Threshold).
+     *
+     * @pre the device is on
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] mode the CCA mode
+     *
+     * @return 0 on success
+     * @return -ENOTSUP if the mode is not supported
+     * @return negative errno on error
+     */
+    int (*set_cca_mode)(ieee802154_dev_t *dev, ieee802154_cca_mode_t mode);
+
+    /**
+     * @brief Set IEEE802.15.4 PHY configuration (channel, TX power)
+     *
+     * This function SHOULD NOT validate the PHY configuration unless
+     * it is specific to the device. The upper layer is responsible of all kind
+     * of validations.
+     *
+     * @pre the device is on
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] conf the PHY configuration
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*config_phy)(ieee802154_dev_t *dev, ieee802154_phy_conf_t *conf);
+
+    /**
+     * @brief Set the transceiver PHY state
+     *
+     * @pre the device is on
+     *
+     * @note the implementation MUST block until the state change occurs.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] state the new state
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_trx_state)(ieee802154_dev_t *dev, ieee802154_trx_state_t state);
+
+```
+
+### Device State Management
+```c
+    /**
+     * @brief Process radio IRQ.
+     *
+     * This function calls the @ref ieee802154_cb_t::cb function with
+     * the corresponding event type.
+     *
+     * @note if the device is off, this function should do nothing
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     */
+    void (*irq_handler)(ieee802154_dev_t *dev);
+
+    /**
+     * @brief Turn on the device
+     *
+     * @pre the init function of the radio succeeded.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     *
+     * @post the transceiver state ins @ref IEEE802154_TRX_STATE_TRX_OFF
```

Comment:
`IEEE802154_TRX_STATE_TRX_ON`?

Reply:
The three states in the RDM are TX_ON, RX_ON, or TRX_OFF. TRX_ON would be a state in which the transceiver would be simultaneously sending while receiving. This is physically very challenging due to the near far effect; but if I recall correctly some research did indeed succeed in filtering out one's own transmission from the input signal, so that the remote signal indeed could be demodulated. But I expect that IoT devices will not have such fancy stuff any time soon. Would be cool enough to have this in our notebook PCs within the next few years :-)